src/MapViewController.swift

```diff
@@ -813,6 +834,81 @@ open class MapViewController: UIViewController, LocationManagerDelegate {
     tgViewController.mapViewDelegate = self
   }
 
+  func reloadTGViewController() {
+    guard let unwrappedSaver = stateSaver else { return }
+    setupTgControllerView()
+    tgViewController.gestureDelegate = self
+    tgViewController.mapViewDelegate = self
+    do {
+      try loadStyleAsync(unwrappedSaver.mapStyle) { [unowned self] (styleLoaded) in
+        self.recreateMap(unwrappedSaver)
+        self.stateSaver = nil
+      }
+    } catch {
+      // In the event Tangram can't do the reload, we're pretty much dead in the water.
+      return
+    }
+  }
+
+  func recreateMap(_ state: StateReclaimer) {
+    let oldAnnotations = self.currentAnnotations
+    self.currentAnnotations = Dictionary()
+    for (annotation, marker) in oldAnnotations {
+      let newMarker = TGMarker.init(mapView: self.tgViewController)
+      newMarker.point = marker.point
+      newMarker.stylingString = marker.stylingString
+      self.currentAnnotations[annotation] = newMarker
+    }
+    if let currentRoute = self.currentRoute {
+      do {
+        try self.display(currentRoute)
+      } catch {
+        //Silently catch this - we may still be able to recover from here sans route
+      }
+    }
+    self.tilt = state.tilt
+    self.rotation = state.rotation
+    self.zoom = state.zoom
+    self.position = state.position
+    self.cameraType = state.cameraType
+    self.setupAttribution()
+    self.setupFindMeButton()
+  }
+
+  override open func viewWillAppear(_ animated: Bool) {
+    super.viewWillAppear(animated)
+    //We only want to attempt to reload incase we reloaded due to memory issues
+    if (!tgViewController.isViewLoaded) {
```

Comment:
I went back and forth on that. My only concern there was if we wind up using state saver for other things but I do agree it would be more resilient not to rely on Tangram. Will update.